### PR TITLE
feat(LOC-2321): Change Tooltip to use React Portals and migrate to functional component

### PR DESCRIPTION
**Pull request by flycrum**
## Summary

This PR is intended to introduce non-breaking changes while introducing React Portals to overcome Tooltips getting clipped by overview containers. In addition, the `Tooltip` component was class-based and should be migrated over to a functional component that uses hooks.

## Technical
- the very nature of adding React Portals changes a number of layout and css rules that might result in the tooltip shifting any number of pixels. This is due to the Tooltip no longer being next to the component in the DOM tree but rather being added as a child to the body element (via portals). These shifts in position have been attempted to be minimized.
- the overall style of the Tooltip should not have changed
- the various states required for the modified Tooltip have changed and there is now an addition delay stage to show the tooltip that results in it not immediately appearing after hovering over the target but rather waiting for a period of time and then checking to see if the user is still hover over the target (or popper) before transitioning in the tooltip. This is configurable via a new `showDelay` prop.
- a `hideDelay`prop has also been added that determines how long the tooltip remains showing after the user's mouse leaves both the target (or popper)
- a new `offsetPopper` prop has been added to expose an internal library option to shift the tooltip popper (if needed)
- the layout and position of the popper arrow had to be reworked but every effort to retain the previous look, feel, size, and position was taken
- a new internal (not exported) `Portal` component has been added that abstracts and streamlines the implementation and allows other components to potentially use it that should lead to consistency
- three new internal hooks have been created that abstracts out logic for detecting hover and transitionEnd along with managing timeouts and swappable timeouts
- `react-popper` was updated to `2.2.4` to leverage to the `usePopper` hook along with `@popperjs/core` being added
- `@storybook/react` was updated to `6.1.11` to support the upgrade to `react-popper` and not conflict with their own internal usage of that package
- fixed storybook example wrapper to set `position: relative` on container div so examples that leverage `position: absolute` doesn't try to anchor to unintended element somewhere else up the DOM tree (this fixed several positioning bugs)
- added stories to `Tooltip` to demonstrate how every possible `position` prop option renders along with showing the new `offsetPopper` and `showDelay` and `hideDelay` props

## Screenshots / Text Samples

image

## Reference

- [LOC-2321](https://getflywheel.atlassian.net/browse/LOC-2321)

### CODE REVIEW

**Review by tihal — COMMENTED** [review]
Reviewed in Storybook, in Local, and in VS Code!

Everything looks like it's working as expected which is great!

Left a couple comments that are mostly nitpicks, but may be small cleanup items.

I think this is a great improvement to our component library - the base implementation of Tooltip is much cleaner and easier to understand now.

After spending some time browsing `useTooltip` and `useTooltipStage`, I will say these two methods feel very complex and intertwined still. It definitely highlights some of the issues using functional components and React hooks with timeouts, etc. I don't have specific recommendations to improve these methods though, and I don't know that the functionality we want can be built without this level of complexity.

Overall amazing effort and work! Excited to get this merged in with FlyTooltip eventually and finally be rid of the pesky css conflicts.

File: src/components/overlays/Tooltip/Tooltip.tsx [hunk 1]
```diff
@@ -2,185 +2,368 @@ import * as React from 'react';
...
+interface ReducerState {
+	stage:
+	| '0Hidden'
+	| '1WaitingToFadeIn'
+	| '2FadingInToShow'
+	| '3Showing'
+	| '4WaitingToFadeOut'
+	| '5FadingOutToHide',
+	direction: ReducerStageDirection,
+}
```
[the remark this answers is not in the document]
Comment (flycrum) [reply src/components/overlays/Tooltip/Tooltip.tsx:39]: Great question! It is not necessary. I've been seeing it more and more in the TS world however as a way of aligning all the values so it's more readable, so I thought I might just adopt it, as well 😄 Thoughts?

File: src/components/overlays/Tooltip/Tooltip.tsx [hunk 2]
```diff
@@ -2,185 +2,368 @@ import * as React from 'react';
...
+		default:
+			throw new Error();
 	}
 
-	protected _onMouseEnterContentPopper = () => {
-		// cancel any timeout attempts to hide tooltip
-		clearTimeout(this._delayTimeoutIdPopper);
-		clearTimeout(this._delayTimeoutIdReference);
+	// do quick and dirty diff check and if nothing has changed then return previous state otherwise return new changed state
+	// return newState.direction === prevState.direction || newState.stage === prevState.stage ? prevState : newState;
```
[the remark this answers is not in the document]
Comment (flycrum) [reply src/components/overlays/Tooltip/Tooltip.tsx:120]: Good call. I'll remove them!

File: src/components/overlays/Tooltip/Tooltip.tsx [hunk 3]
```diff
@@ -2,185 +2,368 @@ import * as React from 'react';
...
+		</>
+	);
 }
+
+Tooltip.defaultProps = {
+	hideDelay: 500,
+	forceHover: false,
+	offsetPopper: [0, 10],
+	position: 'top',
+	showDelay: 1000,
```
[the remark this answers is not in the document]
Comment (flycrum) [reply src/components/overlays/Tooltip/Tooltip.tsx:368]: Trying out other apps on my Mac, I'm seeing quite a wide range for their comparable delay. These are my estimates for some random app samplings:
```
Chrome: 1400
VSCode: 1200 for menu items, 500 for in-code hovers
Fork: 1600
Spotify: 500
Slack: 300
Safari: 1200
```

So, my opinion is that there isn't necessarily a best value for this and it is really open to opinions. My strong preference would be for Tyson to decide how he wants it to feel. It's also kind of interesting cause there are a mixture of uses for us that range from really being the secondary action for something like a menu item (in which case I feel a longer delay is better) vs an info button where it is the primary action to hover over it and (should probably be faster). I wonder if we set the default to slower but when it's the primary action we manually change it to something like `500`. Thoughts?

File: src/components/overlays/Tooltip/Tooltip.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (tihal) [reply src/components/overlays/Tooltip/Tooltip.tsx:39]: Nice! That's good to know, and it _does_ look better. Let's keep it

File: src/components/overlays/Tooltip/Tooltip.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (tihal) [reply src/components/overlays/Tooltip/Tooltip.tsx:368]: Setting to 500 on a primary action I think is a great idea - that was my main concern, was that the delay felt too long and people may not register a component as containing a tooltip at all due to impatience.

Maybe this is something we come back to after merging in FlyTooltip, and we go through and categorize each instance as "primary" or not?

File: src/components/overlays/Tooltip/Tooltip.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (flycrum) [reply src/components/overlays/Tooltip/Tooltip.tsx:39]: Sounds good. Thanks!

**Review by tihal — APPROVED** [review]

File: src/components/overlays/Tooltip/Tooltip.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (flycrum) [reply src/components/overlays/Tooltip/Tooltip.tsx:368]: I really like that idea! 🙌

**Pull request merged**